src/stan_math_backend/Stan_math_code_gen.ml

```diff
@@ -602,6 +632,37 @@ std::vector<double> to_doubles__(std::initializer_list<double> x) {
 std::vector<stan::math::var> to_vars__(std::initializer_list<stan::math::var> x) {
   return x;
 }
+
+inline void validate_positive_index(const char* var_name, const char* expr,
```

Comment:
Would you mind explaining why these new functions were necessary?

Reply:
They're just there for nicer error messages. Stan Math library has only `validate_non_negative_index`. For local variables we just need to check that the declared dimensions are nonnegative but for constrained parameters we must also ensure that the corresponding unconstrained type is valid. An N-dimensional simplex is represented internally by (N-1)-dimensional vector. Calling `validate_non_negative_index` with the calculated unconstrained size would give a confusing error message (the user asked for size 0 simplex so why does it complain about size -1 vector?).

The issue with unit vectors is more subtle. The constraining transform does not change the dimension. However, the transform is discontinuous which causes problems, especially in low dimensions, and the space of _one-dimensional_ unit vectors contains only two points: `[1.0]'` and `[-1.0]'`. HMC cannot possibly sample it efficiently.